### PR TITLE
Fixes chi1_selections (#4108)

### LINKED ISSUE #4108: `chi1_selections` fails for residues with `CG1`
## Expected behavior ##

I expect the single-residue `chi1_selection` and the multi-residue `chi1_selections` (notice the "s") to function identically for a given residue.  Both should recognize `CG1` as the fourth atom of the dihedral.

## Actual behavior ##

The multi-residue `chi1_selections` returns `None` for residues with `CG1`.  But the single-residue `chi1_selection` is correct, thanks to https://github.com/MDAnalysis/mdanalysis/issues/3238.

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD
u = mda.Universe(PSF, DCD)

# This gives the correct result.
print(u.residues[38].chi1_selection())

# This incorrectly returns `None`.
print(u.residues[38:39].chi1_selections())
```

## Current version of MDAnalysis ##

I'm using version MDAnalysis version 2.3.0.  The bug is also present in the source code for 2.5.0-dev0.

### PR DESCRIPTION
**Pull request by DanielJamesEvans**
Fixes #4108

Changes made in this Pull Request:
 - In `chi1_selections`, I made the following changes:
   -  Changed the default value of the function argument `cg_name`
   - Redefined the internal variable `keep`
   - Redefined the internal variable `ags`

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

----
:books: Documentation preview :books:: https://readthedocs-preview--4109.org.readthedocs.build/en/4109/

### CODE REVIEW

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thank you for the fix --- looks solid. 

- [x] please introduce yourself on the developer list (mention your PR) if you haven't done so already
- [x] add a test that check for the new functionality (eg use what you mentioned in your issue report) --- find the current tests for chi1_selections() and add it there

**Comment by DanielJamesEvans** [comment]
Thanks for the reply!  I added a test for the new functionality.  When I pushed this, the CI checks for ubuntu-latest, 3.8 returned an error.  I checked the logs; I'm not sure what happened but it seems to be an error related to CodeCov itself rather than the MDAnalysis update.  The error is this: `Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255`.  Googling it led me to [this issue report on the codecov repo](https://github.com/codecov/codecov-action/issues/598) which doesn't have a conclusive answer.  Do you have any advice?

**Comment by IAlibay** [comment]
> Thanks for the reply! I added a test for the new functionality. When I pushed this, the CI checks for ubuntu-latest, 3.8 returned an error. I checked the logs; I'm not sure what happened but it seems to be an error related to CodeCov itself rather than the MDAnalysis update. The error is this: `Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255`. Googling it led me to [this issue report on the codecov repo](https://github.com/codecov/codecov-action/issues/598) which doesn't have a conclusive answer. Do you have any advice?

Temporary server errors with codecov, I've restarted the job.

**Comment by orbeckst** [comment]
Codecov is flaky and sometimes fails. That's not your fault so just point out that the failure is due to codecov and don't worry about it.

~~I'll kick the runner to rerun, maybe it uploads a bit later.~~ @IAlibay beat me to it.

File: package/MDAnalysis/core/topologyattrs.py [hunk 1]
```diff
@@ -1266,13 +1266,13 @@ def chi1_selections(residues, n_name='N', ca_name='CA', cb_name='CB',
         """
         results = np.array([None]*len(residues))
         names = [n_name, ca_name, cb_name, cg_name]
-        keep = [all(sum(r.atoms.names == n) == 1 for n in names)
+        keep = [all(len(r.atoms.select_atoms(f"name {n}")) == 1
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: Why did you change to using a selection instead of keeping the faster array construct?

Unless there's a good reason, change back. You can time the two approaches and I am sure you'll find that select_atoms is substantially slower.

File: package/MDAnalysis/core/topologyattrs.py [hunk 2]
```diff
@@ -1266,13 +1266,13 @@ def chi1_selections(residues, n_name='N', ca_name='CA', cb_name='CB',
...
-        keep = [all(sum(r.atoms.names == n) == 1 for n in names)
+        keep = [all(len(r.atoms.select_atoms(f"name {n}")) == 1 
+                    for n in names)
                 for r in residues]
         keepix = np.where(keep)[0]
         residues = residues[keep]
 
-        atnames = residues.atoms.names
-        ags = [residues.atoms[atnames == n] for n in names]
+        ags = [residues.atoms.select_atoms(f"name {n}") for n in names]
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: as above — keep existing code unless change is _really_ necessary

**Review by orbeckst — CHANGES_REQUESTED** [review]
Test looks good but I have a question about a potential performance regression, please see comments.

File: package/MDAnalysis/core/topologyattrs.py (hunk 1 above)
Comment (DanielJamesEvans) [reply package/MDAnalysis/core/topologyattrs.py]: The old code fails when `cg_name` includes multiple names.  E.g. if `cg_name='CG CG1 OG OG1 SG'` then the old code checks whether each atom's name is equal to the entire string `'CG CG1 OG OG1 SG'`.  This always returns False.  So changing this line is a key part of fixing the bug.

I based my fix on `chi1_selection`, which also uses atom selection.  But I agree that this is slow.  I wrote an alternate version that avoids atom selection.  In my testing, the newest code takes 18 us while the atom selection code takes 147 us.

Should I submit a PR to change `chi1_selection` to use the faster procedure?  For now, I'm just going to push a version with the newest `chi1_selections`.

File: package/MDAnalysis/core/topologyattrs.py (hunk 2 above)
Comment (DanielJamesEvans) [reply package/MDAnalysis/core/topologyattrs.py]: Like above, the code needs to be changed but I implemented a line that's faster than atom selection.

File: package/MDAnalysis/core/topologyattrs.py (hunk 1 above)
Comment (orbeckst) [reply package/MDAnalysis/core/topologyattrs.py]: Thanks for the explanation.

Please push an update for _your_ `chi1_selections` with faster code.

Let us have a look at it. We can then decide if we want to update chi1_selection right away or in a separate PR.

File: package/MDAnalysis/core/topologyattrs.py [hunk 3]
```diff
@@ -1266,13 +1266,13 @@ def chi1_selections(residues, n_name='N', ca_name='CA', cb_name='CB',
         """
         results = np.array([None]*len(residues))
         names = [n_name, ca_name, cb_name, cg_name]
-        keep = [all(sum(r.atoms.names == n) == 1 for n in names)
-                for r in residues]
+        keep = [all(sum(np.in1d(r.atoms.names, n.split())) == 1
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1270]: I like your updated code.

@IAlibay @richardjgowers are we ok with also updating `chi1_selection()` with the same solution, which avoids `select_atoms()`? We could do a separate PR but this seems unnecessary bureaucracy to me, unless we want a cleaner separation of work in CHANGELOG.

Comment (orbeckst) [reply package/MDAnalysis/core/topologyattrs.py:1270]: Context https://github.com/MDAnalysis/mdanalysis/pull/4109#discussion_r1156624351

> I based my fix on chi1_selection, which also uses atom selection. But I agree that this is slow. I wrote an alternate version that avoids atom selection. In my testing, the newest code takes 18 us while the atom selection code takes 147 us.

Comment (DanielJamesEvans) [reply package/MDAnalysis/core/topologyattrs.py:1270]: I was thinking about the CHANGELOG; I had a couple ideas:

1. List the changes to both `chi1_selections` and `chi1_selection` in a single entry in the Fixes section. Maybe something like this: "_Fix chi1_selections incorrectly returning None (Issue 4108) and chi_selection running unnecessarily slowly_".
2. List the change to `chi1_selections` as a Fix, and the change to `chi1_selection` as an enhancement.   Maybe "_Fix chi1_selections incorrectly returning None (Issue 4108)_" in Fixes, and "_Improved speed of chi1_selection_" in Enhancements.

I think I favor option (2).  But I don't know the history or culture behind this project's CHANGELOG, so I don't have a strong opinion.  And it's definitely possible there's a better option that I haven't thought of!

Comment (orbeckst) [reply package/MDAnalysis/core/topologyattrs.py:1270]: (2) seems better to me

Comment (orbeckst) [reply package/MDAnalysis/core/topologyattrs.py:1270]: @DanielJamesEvans I haven't heard any opposition to you updating `chi1_selection()` in this PR. I encourage you to add it to your PR. Once you've done so, please ping me so that I can review again.

Obviously, anyone else is also welcome to review and if dissenting opinions are voiced in other reviews we can address them then.

Comment (DanielJamesEvans) [reply package/MDAnalysis/core/topologyattrs.py:1270]: Sounds good!  I added the `chi1_selection()` update to my PR.

File: package/CHANGELOG [hunk 4]
```diff
@@ -37,6 +38,7 @@ Fixes
     (Issue #3336)
 
 Enhancements
+  * Improved speed of chi1_selection
```
Comment (orbeckst) [inline package/CHANGELOG]: add PR number

File: package/CHANGELOG [hunk 5]
```diff
@@ -21,6 +21,7 @@ The rules for this file:
  * 2.5.0
 
 Fixes
+  * Fix chi1_selections incorrectly returning None (Issue #4108)
```
Comment (orbeckst) [inline package/CHANGELOG]: On second reading: make it more specific eg "Fix chi1_selections() ignored atom names CG1 OG OG1 SG and incorrectly returned None for amino acids CYS, ILE, SER, THR, VAL (Issue #4108)"

This will tell people immediately if they need this release.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Just minor changes, then it's good to go from my end.

**Review by orbeckst — APPROVED** [review]
Nice work, looks all good to me now.

I’ll wait for a day until merging so that others have a chance to comment.

**Comment by orbeckst** [comment]
Congratulations on your 1st merged PR 🎉 , @DanielJamesEvans !

**Comment by DanielJamesEvans** [comment]
Thanks a lot!  And thanks for the helpful code review!  I look forward to continuing to use (and hopefully occasionally contributing to) this project.
